textpipe/doc.py

```diff
@@ -15,49 +17,63 @@ class Doc:
+    language: 2-letter code for the language of the text
+    is_detected_language: a boolean indicating if the language was specified
+                            beforehand or detected
+    hint_language: language you expect your text to be
     """
 
     def __init__(self, raw, language=None, hint_language='en'):
-        """
-        Args:
-        raw: incoming, unedited text
-        is_detected_language: a boolean indicating if the language was specified
-                              beforehand or detected
-        _language: 2-letter code for the language of the text
-        _hint_language: language you expect your text to be
-        _clean_text: string containing the cleaned text
-        _spacy_nlps: dictionary containing a spacy language module
-        _spacy_doc: dictionary containing an instance of a spacy doc
-        _text_stats: dictionary containing an instance of textacy textstats
-        """
-
         self.raw = raw
         self.is_detected_language = language is None
+        self.hint_language = hint_language
         self._language = language
-        self._hint_language = hint_language
-        self._clean_text = None
-        self._spacy_nlps = {}
-        self._spacy_doc = {}
+
         self._text_stats = {}
 
     @property
     def language(self):
         """
-        Detect the language of a text if no language was provided along with the text
+        Provided or detected language of a text
 
-        >>> doc = Doc('Test sentence for testing text', language='en')
-        >>> doc.language
+        >>> from textpipe.doc import Doc
+        >>> Doc('Test sentence for testing text').language
         'en'
-        >>> doc = Doc('Test sentence for testing text')
-        >>> doc.language
+        >>> Doc('Test sentence for testing text', language='en').language
         'en'
+        >>> Doc('Test', hint_language='nl').language
+        'nl'
         """
         if not self._language:
-            _, _, best_guesses = cld2.detect(self.clean_text, hintLanguage=self._hint_language,
-                                             bestEffort=True)
-            self._language = best_guesses[0][1]
+            self._language = self.detect_language(self.hint_language)
         return self._language
 
+    @functools.lru_cache()
+    def detect_language(self, hint_language=None):
+        """
+        Detected the language of a text if no language was provided along with the text
+
+        Args:
+        hint_language: language you expect your text to be
+
+        Returns:
+        language: 2-letter code for the language of the text
+
+        >>> from textpipe.doc import Doc
+        >>> doc = Doc('Test')
+        >>> doc.language
+        'en'
+        >>> doc.detect_language('nl')
+        'nl'
+        """
+        _, _, best_guesses = cld2.detect(self.clean, hintLanguage=hint_language,
+                                         bestEffort=True)
+        return best_guesses[0][1]
```

Comment:
Will `best_guesses` always have at least one element?

Reply:
I did not change this code, but good idea to make this more robust while I am touching it.

Reply:
Actually the [documentation describes that this is exactly what it gives back, a list of 3 tuples with 4 elements](https://github.com/GregBowyer/cld2-cffi/#documentation). So I am OK with leaving this like it was.